### DISCUSSION TITLE
Is it not possible to map from a `Dictionary<string, object?>`?

### DISCUSSION THREAD

**Discussion by glen-84** [riok/mapperly#1039] [Q&A]
This [is possible](https://docs.automapper.org/en/latest/Dynamic-and-ExpandoObject-Mapping.html?highlight=Dictionary) using AutoMapper.

---

Anything implementing:

```cs
IDictionary<string, object>
IDictionary<string, object?>
IReadOnlyDictionary<string, object>
IReadOnlyDictionary<string, object?>
```

```cs
public sealed class Example
{
    public long Id { get; set; }

    public string Name { get; set; }
}
```

```cs
public sealed class ExampleDictionary : IReadOnlyDictionary<string, object?>
{
    // ...
}
```

```cs
[Mapper]
public static partial class ExampleMapper
{
    public static partial Example ExampleDictionaryToExample(ExampleDictionary dictionary);
}
```

```cs
Example example = new ExampleMapper().ExampleDictionaryToExample(new ExampleDictionary()
{
    { "id", 1 }, // Note the difference in character case.
    { "name", "example" }
});
```

The `object` value may itself be a dictionary, so it should be recursive.

**Comment by TimothyMakkison** [comment]
Hey, thanks for the suggestion. Mapperly currently does not support this feature, I've created an issue to add it.

**Reply by glen-84** [reply]
Thanks, but that seems to be more focussed on `dynamic` than dictionaries?

**Reply by TimothyMakkison** [reply]
Good point, I mainly focused on your link 😅

If one of the riok people like the idea I'll spend more time on it. `Dict -> Type` looks like it will be a fun challenge to implement😄

**Reply by latonz** [reply]
The problem with both is how Mapperly can detect the types and convert them as needed as the types are not known at compile time. For example:
```c#
record A(string Name, int Age, B Nested);
record B(decimal Weight);

partial A Map(Dictionary<string, object> src);
```

How would the generated code look like? IMO the hard part is converting the types which are not known at compile time:
```c#
// this is probably not sufficient:
var target = new A((string)src[nameof(A.Name)], (int)src[nameof(A.Age)], new B((decimal)src[nameof(A.Nested)][nameof(B.Weight)]));

// it would need to more look like this:
var target = new A(MapString(src.GetValueOrDefault(nameof(A.Name)), MapInt(src.GetValueOrDefault(nameof(A.Age)), MapToB(src.GetValueOrDefault(nameof(A.Nested)));

B MapB(object? src)
{
    return src switch
    {
        IReadOnlyDictionary<string, object> srcDict => MapBDict(srcDict),
        _ => throw ...
    };
}

B MapBDict(IReadOnlyDictionary<string, object> src)
{
    // ...
}

string MapString(object? src)
{
    return src switch
    {
        string s => s,
        null => "",
        _ => throw ...
    };
}

int MapInt(int src)
{
    return src switch
    {
        string s => int.Parse(s),
        _ => (int)src,
    }
}
```

Questions:
* How to handle missing, but required values: currently Mapperly knows this at compile time and reports a diagnostic. For this case Mapperly does not know it at compile time.
* How to handle type conversions:
  * What source types should be supported
  * What is the fallback (throw, default value, ...): probably needs to be different depending on the target type and whether it is required

**Reply by glen-84** [reply]
> How to handle missing, but required values: currently Mapperly knows this at compile time and reports a diagnostic. For this case Mapperly does not know it at compile time.

It should throw an exception.

It should also respect `RequiredMappingStrategy` so that you can require all non-ignored properties in `A` and `B` to be filled.

> What source types should be supported

Do you mean when mapping _to_ a dictionary? Can this be done generically, or do you have to add support for each kind separately?

> What is the fallback (throw, default value, ...): probably needs to be different depending on the target type and whether it is required

Yes.

Required -> throw.
Not required -> ignore.

I'm not sure about default values. When are they usually used? When a property has no source, but requires a value to satisfy a constructor parameter or similar?

---

Question:

Can you currently map to and from dictionaries by copying their actual properties? If so, does there need to be an attribute to say "don't copy the properties, use the key/values as properties"?

```cs
[MapKeyValuePairs]
partial A Map(Dictionary<string, object> src);
```

**Reply by latonz** [reply]
> Do you mean when mapping to a dictionary? Can this be done generically, or do you have to add support for each kind separately?

Could you provide an implementation how you would expect Mapperly to generate it based on the following example?
```c#
record A(string Name, int Age, B Nested);
record B(decimal Weight);

partial A Map(Dictionary<string, object> src);
```

**Reply by glen-84** [reply]
That example maps _from_ a dictionary.

Should an example implementation use a constructor or property assignment?

**Reply by latonz** [reply]
Exactly, that's what this discussion is about, isn't it? 🤔 
The existing behaviour in Mapperly is to use property assignments except required otherwise (eg. init only properties or for types without a parameterless ctor, which is the case in the example I provided due to the records).

**Reply by glen-84** [reply]
> Exactly, that's what this discussion is about, isn't it? 🤔

You quoted text referring to the "to" scenario. 🙂

My current use case is the "from" scenario, but it should probably support both.

> ... which is the case in the example I provided due to the records

Oh, right. Can't use property assignment there.

Let me think about the implementation for a bit.

**Reply by glen-84** [reply]
How about something like this, where there's a single `GetArgument` function/method instead of one for each type?

```cs
static A MapA(Dictionary<string, object?> src)
{
    return new A(
        GetArgument<string>(nameof(A.Name), src),
        GetArgument<int>(nameof(A.Age), src),
        MapB(
            GetArgument<Dictionary<string, object?>>(nameof(A.Nested), src)));
}

static B MapB(Dictionary<string, object?> src)
{
    return new B(GetArgument<decimal>(nameof(B.Weight), src));
}

static T GetArgument<T>(string parameterName, Dictionary<string, object?> src)
{
    if (src.TryGetValue(parameterName, out var value))
    {
        if (value is T v)
        {
            return v;
        }
        else
        {
            // Type mismatch -- convert, throw, whatever.
        }
    }

    throw new Exception($"Missing required argument '{parameterName}'.");
}

record A(string Name, int Age, B Nested);
record B(decimal Weight);

var dictionary = new Dictionary<string, object?>()
{
    // allow key/property names to be case-insensitive.
    { "Name", "example" },
    { "Age", 25 },
    { "Nested", new Dictionary<string, object?>() { { "Weight", 100m } } }
};

var a = MapA(dictionary);
```

Note: When calling `MapB` inside `MapA`, it should probably not assume that it's also a dictionary, it could also just be a `B`.

Does this help at all?

**Reply by latonz** [reply]
One of the interesting parts would be the `// Type mismatch -- convert, throw, whatever.`... can you elaborate on this? especially the `convert` part.

> // allow key/property names to be case-insensitive.

This can be fixed by using a key comparer when creating the dictionary.

**Reply by glen-84** [reply]
> can you elaborate on this? especially the convert part

I'm new to Mapperly, but the idea would be for it to behave similarly to when a target property has a different type. How is that handled?

> This can be fixed by using a key comparer when creating the dictionary

- This wouldn't be possible if the dictionary already existed (f.e. was created elsewhere).
- I don't think that this is part of the interface (`I(ReadOnly)Dictionary`), so a specific implementation may not support it.

It might be best to have something similar to `PropertyNameMappingStrategy` that allows you to set the expected case of the source or target keys/properties.

**Reply by latonz** [reply]
The big difference is that the source and the target type is usually known at compile time, therefore Mapperly can provide an implementation depending on these types (eg. using a LINQ conversion, use a parse method, ...). This doesn't work if the type is not known at compile time.

**Reply by glen-84** [reply]
I think there are a few scenarios:

1. (anything) -> new Dictionary

    In this case, there's no target value/type when mapping to a _new_ dictionary, so it would always copy the values as they are.

2. (anything) -> existing Dictionary

    In this case, the property may exist in the target dictionary, but there is no generic type.

    If the property exists, pass it as the `target` here, otherwise just use the source value without conversion.

    Something like this might work for some of the conversions:

    ```cs
    static TTarget Convert<TSource, TTarget>(TSource source, TTarget target)
    {
        // From DateTime.
        if (source is DateTime dateTime)
        {
            // To DateOnly.
            if (typeof(TTarget) == typeof(DateOnly))
            {
                return (TTarget)(object)DateOnly.FromDateTime(dateTime);
            }

            // To TimeOnly.
            if (typeof(TTarget) == typeof(TimeOnly))
            {
                return (TTarget)(object)TimeOnly.FromDateTime(dateTime);
            }
        }

        // From string.
        if (source is string @string)
        {
            // To enum.
            if (typeof(TTarget).IsEnum)
            {
                return (TTarget)Enum.Parse(typeof(TTarget), @string);
            }

            // To parsable.
            // Not sure if this is possible without something like:
            // https://github.com/dotnet/csharplang/discussions/6308
        }

        // ...

        return target;
    }
    ```

3. Dictionary -> Object

    In this case, there's a generic target type, but no generic source type.

    ```cs
    static TTarget? Convert<TTarget>(object source)
    {
        // From DateTime.
        if (source is DateTime dateTime)
        {
            // To DateOnly.
            if (typeof(TTarget) == typeof(DateOnly))
            {
                return (TTarget)(object)DateOnly.FromDateTime(dateTime);
            }

            // To TimeOnly.
            if (typeof(TTarget) == typeof(TimeOnly))
            {
                return (TTarget)(object)TimeOnly.FromDateTime(dateTime);
            }
        }

        // From string.
        if (source is string @string)
        {
            // To enum.
            if (typeof(TTarget).IsEnum)
            {
                return (TTarget)Enum.Parse(typeof(TTarget), @string);
            }

            // To parsable.
            // Not sure if this is possible without something like:
            // https://github.com/dotnet/csharplang/discussions/6308
        }

        // ...

        return default;
    }
    ```

I'm not sure how many of the conversions this will support, or how much additional boxing etc. it may cause. Even if automatic conversions were not supported, direct mapping may still be useful.

**Reply by glen-84** [reply]
**Update:**

I've decided not to move forward with my current use case for this feature. It could still be a useful feature to have, but I don't have a need for it right now. That may change again in the future. Thanks for your consideration.

**Reply by glen-84** [reply]
**Update 2:**

On second thought, I haven't made a final decision regarding the use case that I had in mind*, I'm still trying to determine whether or not it's feasible.

`*` Mapping a Hot Chocolate GraphQL `ObjectResult` (which implements `IReadOnlyDictionary<string, object?>`) to an object graph of entities.

The use case would probably require things like custom mapping for properties with specific names (f.e. map a string key/property named `id` using special code that decodes the value and extracts a numeric value). I'm not sure if this is possible with Mapperly, especially in a generic way (maybe before/after mapping).

Apologies for the noise/indecisiveness.

**Reply by MatthewSteeples** [reply]
We're looking at some similar functionality for a similar reason (Azure.Data.Tables) so may end up trying to implement something unless anyone has any objections?

**Reply by latonz** [reply]
@MatthewSteeples would be happy to see support for this in Mapperly. However, this is not a straight forward task and requires a lot of work.

**Comment by latonz** [comment]
Closing in favor of https://github.com/riok/mapperly/issues/1309.
